File: python/2015/03_perfectly_spherical_houses_in_a_vacuum/aoc201503.py
"""AoC 3, 2015: Perfectly Spherical Houses in a Vacuum"""

# Standard library imports
import pathlib
import sys

DIRECTIONS = {">": (1, 0), "<": (-1, 0), "^": (0, 1), "v": (0, -1)}


def parse(puzzle_input):
    """Parse input"""
    return [DIRECTIONS[arrow] for arrow in puzzle_input]


def visit(moves):
    """Visit houses based on moves"""
    pos_x, pos_y = 0, 0
    houses = {(pos_x, pos_y)}

    for dx, dy in moves:
        pos_x, pos_y = pos_x + dx, pos_y + dy
        houses.add((pos_x, pos_y))

    return houses


def part1(data):
    """Solve part 1"""
    return len(visit(data))


def part2(data):
    """Solve part 2"""
    santa = visit(data[::2])
    robosanta = visit(data[1::2])

    return len(santa | robosanta)


def solve(puzzle_input):
    """Solve the puzzle for the given input"""
    data = parse(puzzle_input)
    solution1 = part1(data)
    solution2 = part2(data)

    return solution1, solution2


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))

File: elixir/lib/2015/03_perfectly_spherical_houses_in_a_vacuum/aoc201503.ex
defmodule AOC2015.Day03 do
  @moduledoc """
  Advent of Code 2015, day 3: Perfectly Spherical Houses in a Vacuum
  """
  require AOC

  @steps %{"^" => {0, 1}, ">" => {1, 0}, "v" => {0, -1}, "<" => {-1, 0}}

  @doc """
  Parse input
  """
  def parse(puzzle_input) do
    puzzle_input |> String.split("", trim: true) |> Enum.map(fn c -> @steps[c] end)
  end

  @doc """
  Solve part 1
  """
  def part1(input) do
    input |> visit() |> Enum.count()
  end

  @doc """
  Solve part 2
  """
  def part2(input) do
    santa = input |> Enum.take_every(2) |> visit()
    robosanta = input |> tl |> Enum.take_every(2) |> visit()
    MapSet.union(santa, robosanta) |> Enum.count()
  end

  @doc """
  Visit houses according to instructions, return positions visited

  ## Example:

      iex> visit([{0, 1}, {1, 0}, {0, -1}, {1, 0}])
      MapSet.new([{0, 0}, {0, 1}, {1, 1}, {1, 0}, {2, 0}])
  """
  def visit(houses) do
    houses
    |> Enum.reduce({MapSet.new([{0, 0}]), 0, 0}, fn {dx, dy}, {pos, x, y} ->
      {MapSet.put(pos, {x + dx, y + dy}), x + dx, y + dy}
    end)
    |> elem(0)
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end
